### PR TITLE
MTD reconstruction: revision of the AODSIM event content

### PR DESCRIPTION
**Pull request by fabiocos**
#### PR description:

Revision of the AODSIM event content of MTD, reducing effectively the output size for PU 200 by ~ 1/3. The ```ValueMap```s kept in the events all refer now to the associated ```generalTrack```, and two more are added to allow the MVA quality flag calculation to be performed without the need of a saved ```recoTrackExtras```.
See [this presentation at the MTD DPG meeting](https://indico.cern.ch/event/1049998/contributions/4411647/attachments/2266668/3848600/MTDEventContent_20210618.pdf) for further details.

#### PR validation:

Test wf. 34634.0 runs smoothly, the update of the output is as expected, and the DQM comparison with the existing workflow does not show differences (as of CMSSW_12_0_0_pre2).

### CODE REVIEW

**Comment by srimanob** [comment]
Thanks @fabiocos

**Comment by fabiocos** [comment]
please test

**Comment by jfernan2** [comment]
+1

File: RecoMTD/Configuration/python/RecoMTD_EventContent_cff.py [hunk 1]
```diff
@@ -1,20 +1,22 @@
...
         'keep *_mtdTrackQualityMVA_*_*')
 )
 
 #RECO content
 RecoMTDRECO = cms.PSet(
-    outputCommands = cms.untracked.vstring() 
+    outputCommands = cms.untracked.vstring()
 )
-RecoMTDRECO.outputCommands.extend(RecoMTDAOD.outputCommands)
+RecoMTDRECO.outputCommands.extend(RecoMTDFEVT.outputCommands)
```
Comment (srimanob) [inline RecoMTD/Configuration/python/RecoMTD_EventContent_cff.py:14]: Should this be done in the opposite way, e.g. define RECO content, and add to FEVT? With this way, one can add content in FEVT-only easier in the future.

Comment (fabiocos) [reply RecoMTD/Configuration/python/RecoMTD_EventContent_cff.py]: I do not see a short term use case like that, but I agree on the general principle

**Comment by fabiocos** [comment]
please test

**Comment by srimanob** [comment]
By the way, @fabiocos, the change in Phase-2 we have seen in comparison is expected, I assume. Could you please confirm?
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_0_X_2021-06-20-2300+1018cd/43578/validateJR.html

- floatedmValueMap_trackExtenderWithMTD_pathLength_RECO_obj
- floatedmValueMap_trackExtenderWithMTD_tmtd_RECO_obj

**Comment by jpata** [comment]
We should have RECO defined as the extension of AOD, and FEVT defined as the extension of RECO (see e.g. https://github.com/cms-sw/cmssw/pull/29571, and thx @perrotta for pointing it out).
Could you please check RecoMTD_EventContent_cff.py and RecoLocalFastTime_EventContent_cff.py?

For the test, perhaps we need to rerun with 23434.21?

**Comment by fabiocos** [comment]
@jpata @perrotta @srimanob the event content definition has been revised according to the suggested logic, already present for ```RecoMTD``` only. This logic is IMO safe for additions, but not bullet-proof for subtractions (i.e. if you remove something from AOD which you want to keep in RECO, as in the case of this PR, you need to manually check anyway each tier, as you would do in the case of independent definitions). For this reason I have repeated the same ```keep``` statements in larger ones. Please clarify whether this is in line with your preferred style.

**Comment by fabiocos** [comment]
@srimanob as you may see in the ```TrackExtenderWithMTD```, some ```ValueMap``` were originally referred to the refitted track, no more saved, and are now associated to the corresponding general track. We have plots in the MTD validation for some of them (e.g. pathlength), and no difference is visible. The other values are input to the MVA quality flag, and also for that no difference is visible in the DQM output. The two values for which the RECO test reports differences are among those affected by the change, as far as I can see.

**Comment by perrotta** [comment]
> @jpata @perrotta @srimanob the event content definition has been revised according to the suggested logic, already present for `RecoMTD` only. This logic is IMO safe for additions, but not bullet-proof for subtractions (i.e. if you remove something from AOD which you want to keep in RECO, as in the case of this PR, you need to manually check anyway each tier, as you would do in the case of independent definitions). For this reason I have repeated the same `keep` statements in larger ones. Please clarify whether this is in line with your preferred style.

@fabiocos in that respect I think it is instead safer to have the definitions encapsulated one into the other, without repetitions. 

The RecoLocalFastTime EventContent is so simple that it is probably not the best example to illustrate it. But if you want to remove (e.g.) `mtdClusters` from all contents, you must remember to do so in all them separately (and therefore check all them). And if you want to add something new you are expected to copy/paste the same line in several places, with all the related risks.

On the other hand, if you want to remove it from the RECO event content, but keep it in the FEVT one. having all them encapsulated as we are suggesting would allow you to simply move the corresponding `keep` statement from one event content to the larger one, without risking to forget doing anything.

**Comment by jpata** [comment]
Thanks for the clarification @perrotta. @fabiocos, why not do as follows in `RecoLocalFastTime_EventContent_cff.py`?

```python
RecoLocalFastTimeAOD = cms.PSet(
    outputCommands = cms.untracked.vstring()
)

#RECO content
RecoLocalFastTimeRECO = cms.PSet(
    outputCommands = cms.untracked.vstring(
        'keep *_mtdRecHits_*_*',
        'keep *_mtdClusters_*_*',
    )
)
RecoLocalFastTimeRECO.outputCommands.extend(RecoLocalFastTimeAOD.outputCommands)

#FEVT
RecoLocalFastTimeFEVT = cms.PSet(
    outputCommands = cms.untracked.vstring(
        'keep *_mtdUncalibratedRecHits_*_*',
        'keep *_mtdTrackingRecHits_*_*',
    )
)
RecoLocalFastTimeFEVT.outputCommands.extend(RecoLocalFastTimeRECO.outputCommands)
```
It seems it would remove `_mtdClusters_` from AOD, but keep them in RECO and FEVT, without explicit repetitions. I'm not completely sure I followed the problem you pointed out about removal, perhaps you can clarify (if still relevant)?

**Comment by jpata** [comment]
test parameters:
  - workflows = 23434.21

**Comment by fabiocos** [comment]
```
15:43 farmui01 815> sdiff -w 300 -b cont3 /gpfs/cms/users/cossutti/Timing/event/std/CMSSW_12_0_0_pre2/work/prova/cont3 
edm::SortedCollection<FTLDataFrameT<BTLDetId,BTLSample,mtdhelpers::BTLRowColDecode>,edm::StrictWeakOrdering<FTLDataFrameT<BTLDetId,BTLSample,mtdhelp    edm::SortedCollection<FTLDataFrameT<BTLDetId,BTLSample,mtdhelpers::BTLRowColDecode>,edm::StrictWeakOrdering<FTLDataFrameT<BTLDetId,BTLSample,mtdhelp
edm::SortedCollection<FTLDataFrameT<ETLDetId,ETLSample,mtdhelpers::ETLRowColDecode>,edm::StrictWeakOrdering<FTLDataFrameT<ETLDetId,ETLSample,mtdhelp    edm::SortedCollection<FTLDataFrameT<ETLDetId,ETLSample,mtdhelpers::ETLRowColDecode>,edm::StrictWeakOrdering<FTLDataFrameT<ETLDetId,ETLSample,mtdhelp
edm::OwnVector<TrackingRecHit,edm::ClonePolicy<TrackingRecHit> >    "trackExtenderWithMTD"      ""                "RECO"                                edm::OwnVector<TrackingRecHit,edm::ClonePolicy<TrackingRecHit> >    "trackExtenderWithMTD"      ""                "RECO"    
edm::SortedCollection<FTLRecHit,edm::StrictWeakOrdering<FTLRecHit> >    "mtdRecHits"                "FTLBarrel"       "RECO"                            edm::SortedCollection<FTLRecHit,edm::StrictWeakOrdering<FTLRecHit> >    "mtdRecHits"                "FTLBarrel"       "RECO"    
edm::SortedCollection<FTLRecHit,edm::StrictWeakOrdering<FTLRecHit> >    "mtdRecHits"                "FTLEndcap"       "RECO"                            edm::SortedCollection<FTLRecHit,edm::StrictWeakOrdering<FTLRecHit> >    "mtdRecHits"                "FTLEndcap"       "RECO"    
edm::SortedCollection<FTLUncalibratedRecHit,edm::StrictWeakOrdering<FTLUncalibratedRecHit> >    "mtdUncalibratedRecHits"    "FTLBarrel"       "RECO"    edm::SortedCollection<FTLUncalibratedRecHit,edm::StrictWeakOrdering<FTLUncalibratedRecHit> >    "mtdUncalibratedRecHits"    "FTLBarrel"       "RECO"
edm::SortedCollection<FTLUncalibratedRecHit,edm::StrictWeakOrdering<FTLUncalibratedRecHit> >    "mtdUncalibratedRecHits"    "FTLEndcap"       "RECO"    edm::SortedCollection<FTLUncalibratedRecHit,edm::StrictWeakOrdering<FTLUncalibratedRecHit> >    "mtdUncalibratedRecHits"    "FTLEndcap"       "RECO"
edm::ValueMap<float>                  "trackExtenderWithMTD"      "btlMatchChi2"    "RECO"                                                              edm::ValueMap<float>                  "trackExtenderWithMTD"      "btlMatchChi2"    "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "btlMatchTimeChi2"   "RECO"                                                           edm::ValueMap<float>                  "trackExtenderWithMTD"      "btlMatchTimeChi2"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "etlMatchChi2"    "RECO"                                                              edm::ValueMap<float>                  "trackExtenderWithMTD"      "etlMatchChi2"    "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "etlMatchTimeChi2"   "RECO"                                                           edm::ValueMap<float>                  "trackExtenderWithMTD"      "etlMatchTimeChi2"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackBeta"   "RECO"                                                           edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackBeta"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackPathLength"   "RECO"                                                     edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackPathLength"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackp"   "RECO"                                                              edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackp"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTracksigmat0"   "RECO"                                                        edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTracksigmat0"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTracksigmatmtd"   "RECO"                                                      edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTracksigmatmtd"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackt0"   "RECO"                                                             edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTrackt0"   "RECO"    
edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTracktmtd"   "RECO"                                                           edm::ValueMap<float>                  "trackExtenderWithMTD"      "generalTracktmtd"   "RECO"    
edm::ValueMap<float>                  "mtdTrackQualityMVA"        "mtdQualMVA"      "RECO"                                                              edm::ValueMap<float>                  "mtdTrackQualityMVA"        "mtdQualMVA"      "RECO"    
                                                                                                                                                     >  edm::ValueMap<float>                  "trackExtenderWithMTD"      "pathLength"      "RECO"    
edm::ValueMap<float>                  "tofPID"                    "probK"           "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "probK"           "RECO"    
edm::ValueMap<float>                  "tofPID"                    "probP"           "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "probP"           "RECO"    
edm::ValueMap<float>                  "tofPID"                    "probPi"          "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "probPi"          "RECO"    
edm::ValueMap<float>                  "tofPID"                    "sigmat0"         "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "sigmat0"         "RECO"    
edm::ValueMap<float>                  "tofPID"                    "sigmat0safe"     "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "sigmat0safe"     "RECO"    
                                                                                                                                                     >  edm::ValueMap<float>                  "trackExtenderWithMTD"      "sigmatmtd"       "RECO"    
edm::ValueMap<float>                  "tofPID"                    "t0"              "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "t0"              "RECO"    
edm::ValueMap<float>                  "tofPID"                    "t0safe"          "RECO"                                                              edm::ValueMap<float>                  "tofPID"                    "t0safe"          "RECO"    
                                                                                                                                                     >  edm::ValueMap<float>                  "trackExtenderWithMTD"      "tmtd"            "RECO"    
edm::ValueMap<int>                    "trackExtenderWithMTD"      "generalTrackassoc"   "RECO"                                                          edm::ValueMap<int>                    "trackExtenderWithMTD"      "generalTrackassoc"   "RECO"    
edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixBarrel"      "RECO"                                                           <
edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixEndcap"      "RECO"                                                           <
edmNew::DetSetVector<FTLCluster>      "mtdClusters"               "FTLBarrel"       "RECO"                                                              edmNew::DetSetVector<FTLCluster>      "mtdClusters"               "FTLBarrel"       "RECO"    
edmNew::DetSetVector<FTLCluster>      "mtdClusters"               "FTLEndcap"       "RECO"                                                              edmNew::DetSetVector<FTLCluster>      "mtdClusters"               "FTLEndcap"       "RECO"    
edmNew::DetSetVector<MTDTrackingRecHit>    "mtdTrackingRecHits"        ""                "RECO"                                                         edmNew::DetSetVector<MTDTrackingRecHit>    "mtdTrackingRecHits"        ""                "RECO"    
vector<reco::Track>                   "trackExtenderWithMTD"      ""                "RECO"                                                              vector<reco::Track>                   "trackExtenderWithMTD"      ""                "RECO"    
vector<reco::TrackExtra>              "trackExtenderWithMTD"      ""                "RECO"                                                              vector<reco::TrackExtra>              "trackExtenderWithMTD"      ""                "RECO"    
15:43 farmui01 816> diff -b cont3 /gpfs/cms/users/cossutti/Timing/event/std/CMSSW_12_0_0_pre2/work/prova/cont3                                                                                                                                                                                                                
19a20
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "pathLength"      "RECO"    
24a26
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "sigmatmtd"       "RECO"    
26a29
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "tmtd"            "RECO"    
28,29d30
< edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixBarrel"      "RECO"    
< edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixEndcap"      "RECO"    
15:44 farmui01 817> diff -b reco_cont3 /gpfs/cms/users/cossutti/Timing/event/std/CMSSW_12_0_0_pre2/work/prova/reco_cont3                                                                                                                                                                                                      
17a18
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "pathLength"      "RECO"    
22a24
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "sigmatmtd"       "RECO"    
24a27
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "tmtd"            "RECO"    
26,27d28
< edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixBarrel"      "RECO"    
< edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixEndcap"      "RECO"    
15:44 farmui01 818> diff -b aod_cont3 /gpfs/cms/users/cossutti/Timing/event/std/CMSSW_12_0_0_pre2/work/prova/aod_cont3                                                                                                                                                                                                        
0a1
> edm::OwnVector<TrackingRecHit,edm::ClonePolicy<TrackingRecHit> >    "trackExtenderWithMTD"      ""                "RECO"    
12a14
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "pathLength"      "RECO"    
17a20
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "sigmatmtd"       "RECO"    
19a23
> edm::ValueMap<float>                  "trackExtenderWithMTD"      "tmtd"            "RECO"    
21,22c25,28
< edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixBarrel"      "RECO"    
< edm::ValueMap<int>                    "trackExtenderWithMTD"      "npixEndcap"      "RECO"    
---
> edmNew::DetSetVector<FTLCluster>      "mtdClusters"               "FTLBarrel"       "RECO"    
> edmNew::DetSetVector<FTLCluster>      "mtdClusters"               "FTLEndcap"       "RECO"    
> vector<reco::Track>                   "trackExtenderWithMTD"      ""                "RECO"    
> vector<reco::TrackExtra>              "trackExtenderWithMTD"      ""                "RECO"    
```

@perrotta @jpata the last version should match your favoured style, the output is what expected (see above). At the end of the day it is mostly a matter of taste.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Here's what we see in 23434.21/step3.root using compareProducts (did it manually now, but `enable profiling` could have been used in retrospect) 

```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
 552634.0 ->         0.0    -552634     OLDO  -3.89     TrackingRecHitsOwned_trackExtenderWithMTD__RECO.
 108233.0 ->         0.0    -108233     OLDO  -0.76     FTLClusteredmNewDetSetVector_mtdClusters_FTLEndcap_RECO.
 755745.0 ->         0.0    -755745     OLDO  -5.32     recoTracks_trackExtenderWithMTD__RECO.
      0.0 ->      2131.8       2132     NEWO   0.01     intedmValueMap_trackExtenderWithMTD_npixBarrel_RECO.
  13092.4 ->         0.0     -13092     OLDO  -0.09     floatedmValueMap_trackExtenderWithMTD_pathLength_RECO.
   5726.1 ->         0.0      -5726     OLDO  -0.04     floatedmValueMap_trackExtenderWithMTD_sigmatmtd_RECO.
 136952.0 ->         0.0    -136952     OLDO  -0.96     FTLClusteredmNewDetSetVector_mtdClusters_FTLBarrel_RECO.
      0.0 ->      4267.0       4267     NEWO   0.03     intedmValueMap_trackExtenderWithMTD_npixEndcap_RECO.
   9693.8 ->         0.0      -9694     OLDO  -0.07     floatedmValueMap_trackExtenderWithMTD_tmtd_RECO.
3084220.0 ->         0.0   -3084220     OLDO  -21.70     recoTrackExtras_trackExtenderWithMTD__RECO.
-------------------------------------------------------------
 14214626 ->     9554754   -4659871           -32.8     ALL BRANCHES
```

This looks in line with the expectation/description.

**Comment by fabiocos** [comment]
@jpata yes, it confirms my test above (BTW, which tool is this compareProducts?)

**Comment by jpata** [comment]
> @jpata yes, it confirms my test above (BTW, which tool is this compareProducts?)

it's this one, used by the bot when profiling: https://github.com/slava77/cms-reco-tools/blob/master/compareProducts.sh

**Comment by fabiocos** [comment]
@jpata thanks, that might be useful as a standalone utility within CMSSW as well IMHO.

**Comment by jpata** [comment]
+reconstruction
- drops MTD content from AOD, while keeping it in RECO & FEVT
- step3.root (AOD) size decreases by 1/3 in the phase2 workflow 23434.21
- we see in the reco checks that trackExtenderWithMTD is dropped as expected

**Comment by jfernan2** [comment]
+1

**Comment by srimanob** [comment]
+Upgrade

**Comment by qliphy** [comment]
+1

**Pull request merged**